### PR TITLE
extend uart interface to allow read with timeout

### PR DESCRIPTION
**Pull request by thetoster**
Signed-off-by: bartlomiej.zarnowski <bartlomiej.zarnowski@imgtec.com>

### CODE REVIEW

**Comment by mtusnio** [comment]
Considering this function duplicates the functionality of the default receive function, except it adds a timeout, I'd say it might be best to just add a timeout to the existing function (with 0 disabling timeouts) and add a define like "UART_TIMEOUT_DEFAULT" with either some reasonable value, and "UART_TIMEOUT_DISABLE" defined as 0.

**Comment by francois-berder** [comment]
For this feature, I thought about a get_timeout/set_timeout functions that would set static variables equal to 0 (with a macro like Michał mentioned) at initialisation.

Michał, the only problem with your solution is that it breaks the existing code and we need to add an extra parameter that will usually be constant in a program. I expect people not to change timeout too often.

**Comment by thetoster** [comment]
Ok, so what changes are expected?

**Comment by francois-berder** [comment]
Add a uart_get_timeout/uart_set_timeout to set the timeout for each uart bus. Change the uart_receive function to take into account the timeout. Do not change the name of the function as this would break existing code.

File: include/letmecreate/core/uart.h [hunk 1]
```diff
@@ -86,14 +88,20 @@ int uart_send(const uint8_t *buffer, uint32_t count);
 int uart_receive(uint8_t *buffer, uint32_t count);
 
 /**
- * @brief Receive some data using current UART device.
+ * @brief Get time after which receive data on uart will be timedout. Each Mikrobus socket have independent timeout, call to this function is correlated with
+ * calls to uart_select_bus(uint8_t)
+ *
+ * @param[in] timeout Timeout in ms after which unction will return, even if not full buffer is read
+ */
+uint32_t uart_get_timeout();
```
Comment (francois-berder-imgtec) [inline include/letmecreate/core/uart.h]: should have a void argument

File: src/core/uart.c [hunk 2]
```diff
@@ -320,23 +300,27 @@ int uart_receive_with_timeout(uint8_t *buffer, uint32_t count, uint32_t timeout)
...
+            FD_ZERO(&set);
+            FD_SET(fds[current_mikrobus_index], &set);
+
+            ret = select(fds[current_mikrobus_index] + 1, &set, NULL, NULL, &tmp_timeout);
+            if (ret == -1) {
+                fprintf(stderr, "uart: Failed to read\n");
+                return -1;
+
+            } else if (ret == 0) {
+                return received_cnt;
```
Comment (francois-berder-imgtec) [inline src/core/uart.c]: I would be great to add an error message like uart: Timeout.

**Pull request merged**